packages/remote-feature-flag-controller/src/utils/user-segmentation-utils.ts

```diff
@@ -0,0 +1,49 @@
+import type { Json } from '@metamask/utils';
+import { v4 as uuidV4 } from 'uuid';
+
+import type { FeatureFlagScopeValue } from '../remote-feature-flag-controller-types';
+
+/* eslint-disable no-bitwise */
+/**
+ * Generates a deterministic random number between 0 and 1 based on a metaMetricsId.
+ * This is useful for A/B testing and feature flag rollouts where we want
+ * consistent group assignment for the same user.
+ *
+ * @param metaMetricsId - The unique identifier used to generate the deterministic random number
+ * @returns A number between 0 and 1 that is deterministic for the given metaMetricsId
+ */
+export function generateDeterministicRandomNumber(
+  metaMetricsId: string,
+): number {
+  const hash = [...metaMetricsId].reduce((acc, char) => {
+    const chr = char.charCodeAt(0);
+    return ((acc << 5) - acc + chr) | 0;
+  }, 0);
+
+  return (hash >>> 0) / 0xffffffff;
```

Comment:
We should be careful here about the expected size of the `metaMetricsId`. If the number we're dividing by here isn't the max size of the ID, we'll get an uneven distribution.

i.e. if our target threshold is 10% and `0xffffffff` is double the max hash, then this will result in only 5% of users getting the feature. But if `0xffffffff` is 1/10 of the max hash, it would result in ~99% of users having the feature enabled.

Correct distribution here is dependent upon each of the constants we're using here being exactly correct. e.g. the `5` bit shift above, and the `0xffffffff` here. Ideally we'd assert in code that the max here is truly the max ID (using the length of the ID)

Reply:
What's especially tricky is that the exact length/contents of the ID might differ between MetaMask versions or clients. Maybe not, but I haven't verified that. That could confuse things even more.

Reply:
Ideas for validation:
* Assertion on metametrics length, and derive constants from expected length
* Test cases for minimum and maximum ID (should get values 0 and 1 respectively )
* Validate that all production metametrics IDs in both clients are of the expected length

Reply:
> What's especially tricky is that the exact length/contents of the ID might differ between MetaMask versions or clients. Maybe not, but I haven't verified that. That could confuse things even more.
I had a discussion with @NicolasMassart on this topic, and we decided to go for uuidV4 on this one. 
Currently for mobile side, `metaMetricsId` is uuidv4 format so we expect 36-character string, while in extension side is a hash hex with 66-characters string but it seems to be planned to migrate to uuidv4